### PR TITLE
contracts: Update MultiSigStateMachine

### PR DESCRIPTION
**Pull request by j-mueller**
* Use the new contracts API in `MultiSigStateMachine`
* It now gets the current state of the contract from the chain (an improvement over the old version), but doing this is still a bit awkward
* Tests and error handling are also better now
* I tried to use the state machine & typed transaction libs where possible but I think what we really need is a `TypedUnbalancedTx`

### CODE REVIEW

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs [hunk 1]
```diff
@@ -235,112 +252,85 @@ machineInstance params =
...
+    let addr = Scripts.scriptAddress $ scriptInstance params
+    utxos <- utxoAt addr
+    txouttx <-
+        case (utxos ^. at addr) >>= listToMaybe . Map.toList of
+            Nothing      -> throwing_ _MSContractStateNotFound
+            Just (_, o)  -> pure o
+    Typed.TypedScriptTxOut{Typed.tyTxOutData=state,Typed.tyTxOutTxOut} <- 
+            withContractError (review _MSTypedAPIError)
+            $ Typed.typeScriptTxOut (scriptInstance params) txouttx
+    pure (state, Tx.txOutValue tyTxOutTxOut)
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs]: Although then we're getting into implementing the state machine contract lib I want to write ;)

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs [hunk 2]
```diff
@@ -235,112 +252,85 @@ machineInstance params =
...
-    then do
-        (scriptTx, newState) <- SM.mkHalt (machineInstance prms) currentState Pay
-        addOutAndPay scriptTx
-        pure newState
-    else do
-        (scriptTx, newState) <- SM.mkStep (machineInstance prms) currentState Pay (const valueLeft)
-        addOutAndPay scriptTx
-        pure newState
+        ins       = Typed.collectFromScript utxos (scriptInstance params) Pay
+        outs = if Value.isZero valueLeft
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs]: This needs some comments.

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs [hunk 3]
```diff
@@ -235,112 +252,85 @@ machineInstance params =
...
-    -> m State
-makePayment prms currentState = do
+makePaymentC :: AsMultiSigError e => Params -> Contract MultiSigSchema e ()
+makePaymentC params = do
+    () <- endpoint @"pay"
+    (state, _) <- currentState params
     -- we can't use 'runStep' because the outputs of the transaction are
     -- different from the other transitions: We need two outputs, a public
     -- key output with the payment, and the script output with the remaining
     -- funds.
```
Comment (michaelpj) [inline plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs]: Could use `mkStep` or `mkHalt`?

**Comment by j-mueller** [comment]
Sorry, I merged this one prematurely, I'll address the comments in a follow-up

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs (hunk 3 above)
Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs]: `mkStep` and `mkHalt` both have a `WalletAPI` constraint

Comment (michaelpj) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs]: Yeah, they just need to be changed to take the `AddressMap` as an argument, which is how I've been gradually removing the `WalletAPI` constraints from everything ;)

File: plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSigStateMachine.hs]: I'm going to do https://github.com/input-output-hk/plutus/issues/1384 next which should address this

**Pull request merged**